lib/CliftTransforms/TrivialReturnElimination.cpp

```diff
@@ -0,0 +1,67 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "revng/Clift/Clift.h"
+#include "revng/CliftTransforms/Passes.h"
+#include "revng/CliftTransforms/RewriteHelpers.h"
+
+namespace mlir {
+namespace clift {
+#define GEN_PASS_DEF_CLIFTTRIVIALRETURNELIMINATION
+#include "revng/CliftTransforms/Passes.h.inc"
+} // namespace clift
+} // namespace mlir
+
+namespace clift = mlir::clift;
+using namespace clift;
+
+namespace {
+
+struct TrivialReturnEliminationPattern : mlir::OpRewritePattern<ReturnOp> {
+
+  using OpRewritePattern::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(ReturnOp Return,
+                  mlir::PatternRewriter &Rewriter) const override {
+    if (not Return.getResult().empty())
+      return mlir::failure();
+
+    auto Function = Return->getParentOfType<FunctionOp>();
+    auto JumpTarget = BlockPosition::getEnd(Function.getBody());
+    auto FallTarget = getFallthroughTarget(BlockPosition::getNext(Return));
+
+    if (JumpTarget != FallTarget)
+      return mlir::failure();
+
+    Rewriter.eraseOp(Return);
+    return mlir::success();
+  }
+};
+
+template<typename T>
+using PassBase = clift::impl::CliftTrivialReturnEliminationBase<T>;
+
+struct TrivialReturnEliminationPass : PassBase<TrivialReturnEliminationPass> {
+  void runOnOperation() override {
+    mlir::MLIRContext *Context = &getContext();
+
+    mlir::RewritePatternSet Patterns(Context);
+    Patterns.add<TrivialReturnEliminationPattern>(Context);
+
+    // TODO: Use walkAndApplyPatterns
```

Comment:
There's plenty of this `TODO`s around the codebase. I assume `walkAndApplyPatterns` is an API available upstream that we don't have available in our current LLVM version? If so, feel free to close this thread.